src/environment/environment.cpp

```diff
@@ -83,22 +86,65 @@ Environment::Environment(int argc, char **argv,const char* help,const long int&
 
   PrintStatus();
   GA_Initialize();
+  p_from_comm = true;
   MA_init(C_DBL,ma_stack,ma_heap);
   gridpack::math::Initialize(&argc,&argv);
 
 }
 
+Environment::Environment(int argc, char **argv, MPI_Comm &comm): p_boostEnv(argc,argv),clparser(argc,argv),p_from_comm(false)
+{
+#ifdef ENABLE_ENVIRONMENT_FROM_COMM
+  PrintStatus();
+  pma_stack = 200000;
+  pma_heap  = 200000;
+  if (GA_Initialize_comm(comm)) {
+    p_from_comm = true;
+    MA_init(C_DBL,pma_stack,pma_heap);
+    gridpack::math::Initialize(&argc,&argv);
+  } else {
+    int rank;
+    MPI_Comm_rank(comm,&rank);
+    if (rank == 0) {
+      std::cout<<"Initialize GA from communicator failed" << std::endl;
+    }
+    MPI_Abort(comm, 1);
+  }
+#else
+  int rank;
+  MPI_Comm_rank(comm,&rank);
+  if (rank == 0) {
+    std::cout<<"Initialize GridPACK Environment from communicator not implemented"
+      << std::endl;
+  }
+  MPI_Abort(comm, 1);
+#endif
+}
+
 Environment::~Environment(void)
 {
-  // Finalize math libraries
-  gridpack::math::Finalize();
+  if (p_from_comm) {
```

Comment:
Note the artifacts from rebase.
I don't understand the destructor.  It seems to me that `gridpack::math::Finalize();` and `GA_Terminate()` should be called no matter the state of `p_from_comm`.